### ISSUE TITLE
`unattended-upgrade` with locally edited `routinator.conf` causes "Not found in archive" tar error

### ISSUE THREAD

**Issue by ximon18** [NLnetLabs/routinator#783]
**Note:** This may affect other Routinator versions and other operating systems, but for now the issue is known to affect doing an unattended upgrade of Routinator from 0.11.2 to 0.11.3 on Debian 11.5.

```shell
# unattended-upgrade
...
tar: ./etc/routinator/routinator.conf: Not found in archive
tar: Exiting with failure status due to previous errors
```

Credit to @appliedprivacy for the original problem report and for supplying `unattended-upgrade`  debug and verbose logging output to assist with diagnosing the issue!

**Comment by ximon18** [comment]
I looked into this and found that it is rather surprising, to me at least.

We use a tool called [`cargo-deb`](https://crates.io/crates/cargo-deb) to create the DEB archive. It in turn uses a Rust crate called [`tar`](https://crates.io/tar) to create the inner TAR archive. It uses a function when adding a file to the TAR archive called `set_path()` which notes in [its documentation](https://docs.rs/tar/0.4.38/tar/struct.Header.html#method.set_path
):

> "Will strip out any “.” path component, which signifies the current directory."

Yet `unattended-upgrades` insists on this leading `./` being present in the archive, and other DEB archives do indeed seem to have paths starting with `./` inside their data TAR archive (e.g. I checked the `gpg-agent.deb` archive),

So the problem is two-fold: 
1. How should `cargo-deb` behave, e.g. if the developer invoking cargo-deb doesn't provide `./` prefixed paths should `cargo-deb` force them to be ./ prefixed, and
2. How can `cargo-deb` even add such paths to the TAR archive given the behaviour noted above.

## More details

Enabling debug and verbose logging with `unattended-upgrade` yields the following log fragment:
```shell
check_conffile_prompt(/var/cache/apt/archives/routinator_0.11.3-1bullseye_amd64.deb)
found pkg: routinator
conffile line: /etc/routinator/routinator.conf 
a2fc42b04dc36bb109918396907fd4f7
current md5: 7598e9b7447ed3a1310ecb004bb1e155
tar: ./etc/routinator/routinator.conf: Not found in archive
tar: Exiting with failure status due to previous errors
pkg_md5sum: d41d8cd98f00b204e9800998ecf8427e
Package routinator has conffile prompt and needs to be upgraded manually
Package routinator has conffile prompt and needs to be upgraded manually
Packages blacklist due to conffile prompts: ['routinator$']
Applying pinning: PkgPin(pkg='/^routinator$/', priority=-32768)
Checking the black and whitelist: routinator
skipping blacklisted package routinator
pkg routinator package has been blacklisted
```

The `conffile line` log line comes from a [line](https://github.com/Jimdo/unattended-upgrades/blob/master/unattended-upgrade#L448) inside the [conffile_prompt()](https://github.com/Jimdo/unattended-upgrades/blob/master/unattended-upgrade#L477) function which later calls function [`get_md5_sum_for_file()`}(https://github.com/Jimdo/unattended-upgrades/blob/master/unattended-upgrade#L404), which looks like this:

```python
def get_md5sum_for_file_in_deb(deb_file, conf_file):
    dpkg_cmd = ["dpkg-deb","--fsys-tarfile", deb_file]
    tar_cmd = ["tar","-x","-O", "-f","-", "."+conf_file]
    md5_cmd = ["md5sum"]
    dpkg_p = Popen(dpkg_cmd, stdout=PIPE)
    tar_p = Popen(tar_cmd, stdin=dpkg_p.stdout, stdout=PIPE)
    md5_p = Popen(md5_cmd, stdin=tar_p.stdout, stdout=PIPE)
    pkg_md5sum = md5_p.communicate()[0].split()[0]
    return pkg_md5sum
```

That is equivalent to this:

```shell
$ dpkg-deb --fsys-tarfile routinator_0.11.3-1bullseye_amd64.deb | \
  tar -x -O -f - ./etc/routinator/routinator.conf
tar: ./etc/routinator/routinator.conf: Not found in archive
tar: Exiting with failure status due to previous errors
```

Which is the error noted in the initial report above.

## Comparing our package to others

Note the leading ./ in the following example:
```shell
root@ximon-routinator-unattended-upgrade-test:~# dpkg-deb --fsys-
tarfile /var/cache/apt/archives/gpg-agent_2.2.27-2+deb11u1_amd64.deb |
tar tf - | head -n 3
./
./etc/
./etc/X11/
```

And how it is missing in the routinator archive:
```shell
root@ximon-routinator-unattended-upgrade-test:~# dpkg-deb --fsys-
tarfile /var/cache/apt/archives/routinator_0.11.3-1bullseye_amd64.deb |
tar tf - | head -n 3
usr/
usr/share/
usr/share/doc/
```

## A possible workaround

Once the DEB archive has been built, using the `ar` and `tar` tools extract the `data` tar archive from within the DEB and extract it, then recreate it using leading `./` paths and use the `ar` tool to replace the existing data archive with the updated one. Doing this then passes the `dpkg-deb --fsys-tarfile` test above.

**Comment by ximon18** [comment]
## Workaround proof of concept

```shell
$ docker run --rm -it debian:11
root@a0fe135aa846:/# apt install -y binutils xz-utils
root@a0fe135aa846:/# cd /tmp
root@a0fe135aa846:/tmp# apt download routinator
root@a0fe135aa846:/tmp# dpkg-deb --fsys-tarfile routinator_0.11.3-1bullseye_amd64.deb | tar -x -O -f - ./etc/routinator/routinator.conf
tar: ./etc/routinator/routinator.conf: Not found in archive
tar: Exiting with failure status due to previous errors
root@a0fe135aa846:/tmp# ar x routinator_0.11.3-1bullseye_amd64.deb data.tar.xz
root@a0fe135aa846:/tmp# mkdir t
root@a0fe135aa846:/tmp# tar -C t -xf data.tar.xz
root@a0fe135aa846:/tmp# pushd t
root@a0fe135aa846:/tmp/t# tar cJf ../data.tar.xz ./*
root@a0fe135aa846:/tmp/t# popd
root@a0fe135aa846:/tmp# ar r routinator_0.11.3-1bullseye_amd64.deb data.tar.xz
root@a0fe135aa846:/tmp# dpkg-deb --fsys-tarfile routinator_0.11.3-1bullseye_amd64.deb | tar -x -O -f - ./etc/routinator/routinator.conf | head -n 3
# Configuration for Running Routinator as a System Service
# ========================================================
#
```

Dada!

**Comment by ximon18** [comment]
This should be worked around in the [NLnet Labs resuable packaging workflow](https://github.com/NLnetLabs/.github/tree/main/docs) as the issue is not specific to Routinator.

**Comment by ximon18** [comment]
Git blame on the `get_md5sum_for_file_in_deb()` function/file in `unattended-upgrade` is fun: `initial commit 10 years ago`...

**Comment by ximon18** [comment]
Release [v1.0.4](https://github.com/NLnetLabs/.github/releases/tag/v1.0.4) of the reusable packaging workflow has just been made which includes the workaround for the tar error in this issue.

Once Routinator PR #780 is merged the next release of Routinator will use the reusable packaging workflow and thus include this fix.

**Comment by ximon18** [comment]
I've raised https://github.com/kornelski/cargo-deb/issues/47to discuss this with the upstream authors.

**Comment by ximon18** [comment]
Fixed by PR #780.
